File: backend/src/test/java/com/team05/petmeeting/domain/animal/controller/AnimalControllerTest.java
package com.team05.petmeeting.domain.animal.controller;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.BDDMockito.given;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import com.team05.petmeeting.domain.animal.dto.AnimalRes;
import com.team05.petmeeting.domain.animal.entity.Animal;
import com.team05.petmeeting.domain.animal.service.AnimalService;
import com.team05.petmeeting.global.security.filter.JwtAuthenticationFilter;
import com.team05.petmeeting.global.security.util.JwtUtil;
import java.time.LocalDate;
import java.util.List;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc;
import org.springframework.boot.webmvc.test.autoconfigure.WebMvcTest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.web.servlet.MockMvc;

@WebMvcTest(AnimalController.class)
@AutoConfigureMockMvc(addFilters = false)
@ActiveProfiles("test")
class AnimalControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @MockitoBean
    private AnimalService animalService;

    @MockitoBean
    private JwtUtil jwtUtil;

    @MockitoBean
    private JwtAuthenticationFilter jwtAuthenticationFilter;

    @Test
    @DisplayName("유기동물 리스트 조회 테스트")
    void animalListTest() throws Exception {
        // given
        Animal animal = Animal.builder()
                .desertionNo("12345")
                .processState("보호중")
                .stateGroup(0)
                .noticeNo("공고-1")
                .noticeEdt(LocalDate.now())
                .upKindNm("개")
                .kindFullNm("[개] 믹스견")
                .colorCd("흰색")
                .age("2023(년생)")
                .weight("5kg")
                .sexCd("M")
                .totalCheerCount(10)
                .build();

        AnimalRes animalRes = new AnimalRes(animal);
        Pageable pageable = PageRequest.of(0, 12);
        Page<AnimalRes> page = new PageImpl<>(List.of(animalRes), pageable, 1);

        given(animalService.getAnimals(anyString(), anyString(), anyInt(), any(Pageable.class))).willReturn(page);

        // when & then
        mockMvc.perform(get("/api/v1/animals")
                        .param("region", "서울")
                        .param("kind", "개")
                        .param("stateGroup", "0")
                        .param("page", "0")
                        .param("size", "12"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.content[0].desertionNo").value("12345"))
                .andExpect(jsonPath("$.content[0].kindFullNm").value("[개] 믹스견"))
                .andExpect(jsonPath("$.totalElements").value(1))
                .andDo(print());
    }

    @Test
    @DisplayName("유기동물 상세 조회 테스트")
    void animalDetailTest() throws Exception {
        // given
        Long animalId = 1L;
        Animal animal = Animal.builder()
                .desertionNo("12345")
                .processState("보호중")
                .stateGroup(0)
                .noticeNo("공고-1")
                .noticeEdt(LocalDate.now())
                .upKindNm("개")
                .kindFullNm("[개] 믹스견")
                .colorCd("흰색")
                .age("2023(년생)")
                .weight("5kg")
                .sexCd("M")
                .totalCheerCount(10)
                .build();

        given(animalService.findByAnimalId(eq(animalId))).willReturn(animal);

        // when & then
        mockMvc.perform(get("/api/v1/animals/{animalId}", animalId))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.desertionNo").value("12345"))
                .andExpect(jsonPath("$.kindFullNm").value("[개] 믹스견"))
                .andExpect(jsonPath("$.totalCheerCount").value(10))
                .andDo(print());
    }
}

File: backend/src/test/java/com/team05/petmeeting/domain/animal/controller/AnimalControllerTest.kt
package com.team05.petmeeting.domain.animal.controller

import com.team05.petmeeting.domain.animal.dto.AnimalRes
import com.team05.petmeeting.domain.animal.entity.Animal
import com.team05.petmeeting.domain.animal.entity.Animal.Companion.builder
import com.team05.petmeeting.domain.animal.service.AnimalService
import com.team05.petmeeting.global.security.filter.JwtAuthenticationFilter
import com.team05.petmeeting.global.security.util.JwtUtil
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.mockito.BDDMockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc
import org.springframework.boot.webmvc.test.autoconfigure.WebMvcTest
import org.springframework.data.domain.*
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.bean.override.mockito.MockitoBean
import org.springframework.test.util.ReflectionTestUtils
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultHandlers
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import java.time.LocalDate

@WebMvcTest(AnimalController::class)
@AutoConfigureMockMvc(addFilters = false)
@ActiveProfiles("test")
internal class AnimalControllerTest {
    @Autowired
    lateinit var mockMvc: MockMvc

    @MockitoBean
    lateinit var animalService: AnimalService

    @MockitoBean
    lateinit var jwtUtil: JwtUtil

    @MockitoBean
    lateinit var jwtAuthenticationFilter: JwtAuthenticationFilter

    // given
    val animal = builder()
        .desertionNo("12345")
        .processState("보호중")
        .stateGroup(0)
        .noticeNo("공고-1")
        .noticeEdt(LocalDate.now())
        .upKindNm("개")
        .kindFullNm("[개] 믹스견")
        .colorCd("흰색")
        .age("2023(년생)")
        .weight("5kg")
        .sexCd("M")
        .totalCheerCount(10)
        .build().apply {
            // 💡 범위 지정 함수 apply를 사용해서 객체 생성과 동시에
            // ReflectionTestUtils가 실행되도록 묶어주면 컴파일러의 널 추론 범위가 안전해집니다.
            ReflectionTestUtils.setField(this, "id", 1L)
        }

    @Test
    @DisplayName("유기동물 리스트 조회 테스트")
    @Throws(Exception::class)
    fun animalListTest() {
        // given
        val animalRes = AnimalRes(animal)

        // 컨트롤러의 @PageableDefault(sort = ["noticeEdt"], direction = Sort.Direction.ASC)와 일치시켜야함
        val pageable: Pageable = PageRequest.of(
            0,
            12,
            Sort.by(Sort.Direction.ASC, "noticeEdt")
        )
        val page: Page<AnimalRes> = PageImpl(listOf(animalRes), pageable, 1)

        BDDMockito.given<Page<AnimalRes>>(
            animalService.getAnimals(
                "서울",
                "개",
                0,
                pageable
                )
            ).willReturn(page)

        // when & then
        mockMvc.perform(
            MockMvcRequestBuilders.get("/api/v1/animals")
                .param("region", "서울")
                .param("kind", "개")
                .param("stateGroup", "0")
                .param("page", "0")
                .param("size", "12")
        )
            .andExpect(MockMvcResultMatchers.status().isOk())
            .andExpect(MockMvcResultMatchers.jsonPath("$.content[0].desertionNo").value("12345"))
            .andExpect(MockMvcResultMatchers.jsonPath("$.content[0].kindFullNm").value("[개] 믹스견"))
            .andExpect(MockMvcResultMatchers.jsonPath("$.totalElements").value(1))
            .andDo(MockMvcResultHandlers.print())
    }

    @Test
    @DisplayName("유기동물 상세 조회 테스트")
    @Throws(Exception::class)
    fun animalDetailTest() {
        // given
        val animalId = 1L

        BDDMockito.given<Animal>(animalService.findByAnimalId(animalId))
            .willReturn(animal)

        // when & then
        mockMvc.perform(MockMvcRequestBuilders.get("/api/v1/animals/{animalId}", animalId))
            .andExpect(MockMvcResultMatchers.status().isOk())
            .andExpect(MockMvcResultMatchers.jsonPath("$.desertionNo").value("12345"))
            .andExpect(MockMvcResultMatchers.jsonPath("$.kindFullNm").value("[개] 믹스견"))
            .andExpect(MockMvcResultMatchers.jsonPath("$.totalCheerCount").value(10))
            .andDo(MockMvcResultHandlers.print())
    }
}
